design/kata-2-0-metrics.md

```diff
@@ -0,0 +1,1343 @@
+
+The way `magent` get shim's metrics socket file(`magent_address`) like that `containerd` get shim address. The socket is an abstract socket and saved as file `abstract` with the same directory of `address` for `containerd`.
+
+**Note**: If there is no Prometheus server is configurated, i.e., there is no scrape operations, `magent` will do nothing initiative.
+
+### Kata runtime
+
+Runtime are mainly to:
+
+- Gather metrics about shim process
+- Gather metrics about hypervisor process
+- Gather metrics about running sandbox
+- Get metrics from Kata agent(through `ttrpc`)
+
+### Kata agent
+
+Agent is responsible for:
+
+- Gather agent process metrics
+- Gather guest OS metrics
+
+And in Kata 2.0, agent will add a new interface:
+
+```
+rpc GetMetrics(GetMetricsRequest) returns (Metrics);
+
+message GetMetricsRequest {}
+
+message Metrics {
+	string metrics = 1;
+}
+
+```
+
+The `metrics` field is Prometheus encoded content. This can avoid defining a fixed struct in protocol buffers.
+
+### Performance and overhead
+
+Metrics should not become the bottleneck of system, downgrade the performance, and run with minimal overhead.
+
+|\*|No Sandbox | 1 Sandbox | 2 Sandboxes |
+|---|---|---|---|
+|Metrics count| 39 | 106 | 173 |
+|Metrics size(bytes)| 9K | 144K | 283K |
+|Metrics size(gzipped, bytes)| 2K | 10K | 17K |
+
+*Metrice size*: Response size of one Prometheus scrape request.
+
+It's easy to estimated that if there are 10 sandboxes running in the host, the size of one metrics fetch request issued by Prometheus will be about to 9 + (144 - 9) * 10 = 1.35M (un-gzipped) or 2 + (10 - 2) * 10 = 82K (gzipped). Of course Prometheus support gzip compression, that can reduce the response size of every request.
+
+And here is some test data:
+
+- E2E(from request to `magent` and `magent` write response back): 20ms(avg)
+- Agent(RPC all from shim to agent): 3ms(avg)
+
+Test infrastructure:
+
+- OS: Ubuntu 20.04
+- Hardware: Intel(R) Core(TM) i5-8500 CPU @ 3.00GHz, 6 Cores, and 16GB memory.
+
+**Scrape interval**
+
+Prometheus default `scrape_interval` is 1 minute, and usally it is set to 15s. Small `scrape_interval` will cause more overhead, so user should set it on monitoring's demand.
+
+## Metrics list
+
+Here listed is all supported metrics by Kata 2.0. Some metrics is depended on guest kernels in the VM, so there may be some different by your environment.
+
+Metrics is categorized by component where metrics are collected from and for.
+
+* [Kata agent metrics](#kata-agent-metrics)
+* [Kata guest OS](#kata-guest-os)
+* [Hypervisor metrics](#hypervisor-metrics)
+* [Kata magent metrics](#kata-magent-metrics)
+* [Kata shim metrics](#kata-shim-metrics)
+
+**NOTE**: Labels here are not include `instance` and `job` label that added by Prometheus.
+
+### Kata agent metrics
+
+Agent's metrics contains metrics about agent process.
+
+
+#### kata_agent_proc_stat (GAUGE)
+
+Agent process stat.
```

Comment:
I'd suggest putting all these metrics into tables. Something like:

| Metrics provider | Metric name | Units | Description | Introduced in Kata version |
|-|-|-|-|-|
| Kata Agent | `cstime` | seconds | blah, blah, blah | 2.0.0 |
| Kata Runtime | `foos_per_second` | foo calls / second | blah, blah, blah | 2.1.0 |